malcolm/modules/pmac/velocityprofile.py

```diff
@@ -0,0 +1,451 @@
+        quantization). Otherwise the combined profile may contain very small
+        time intervals that pmac PVT cannot achieve.
+
+        Returns bool: true if this profile requires quantization:
+        """
+
+        def is_quantized(val  # type: float
+                         ):
+            decimals = val % 1 * 1000
+            return np.isclose(decimals, np.round(decimals))
+
+        times = np.array([self.t1, self.tm, self.t2])
+        result = not np.vectorize(is_quantized)(times).all()
+        return result
+
+    def quantize(self):
+        """
+        ensure that all time points are exactly on 1 millisecond boundaries
+        do this by:
+            add 1 milliseconds to t1, tm, t2 and round down
+            adjust vm downwards so that d is correct
+        When reducing vm, keep t1, tm the same but adjust the
+        acceleration downwards.
+
+        the function for vm was derived by taking the sum of the area of
+        two trapezoids and a rectangle described by v1, vm, v2, t1, tm, t2
+        then solving for vm
+        https://www.wolframalpha.com/input/?i=solve+d%3D%28v1%2Bv0%29%2F2
+        +t1+%2B+v0+t0+%2B+%28v2%2Bv0%29%2F2+t2+for+v0
+
+        :Returns Array(float), Array(float): absolute time and velocity arrays
+        """
+
+        # round up the two acceleration times t1, t2 to the nearest millisecond
+        # then round up the middle constant velocity tm such that the total
+        # time tv2 is tv2 + 2 milliseconds rounded up
+        # This approach preserves symmetry but at the same time ensures that
+        # total time is increased deterministically plus acceleration and
+        # vm are decreased
+        self.t1 = np.ceil(self.t1 * 1000) / 1000
+        self.t2 = np.ceil(self.t2 * 1000) / 1000
+        self.tv2 = np.ceil(self.tv2 * 1000 + 2) / 1000
+        self.tm = self.tv2 - self.t1 - self.t2
+
+        i1 = -2 * self.d + self.t1 * self.v1 + self.t2 * self.v2
+        i2 = 2 * self.tm + self.t1 + self.t2
+        self.vm = - i1 / i2
+        self.quantized = True
+
+    def make_arrays(self):
+        """
+        Convert the time and velocity properties to arrays for consumption
+        by pmac/util.py
+
+        :Returns Array(float), Array(float): absolute time, velocity arrays
+        """
+        # return ABSOLUTE time and velocity arrays to describe the profile
+        time_array = [0.0, self.t1, self.tv2]
+        velocity_array = [self.v1, self.vm, self.v2]
+        if self.tm > 0:
+            time_array.insert(2, self.t1 + self.tm)
+            velocity_array.insert(2, self.vm)
+        if self.settle_time > 0:
+            time_array.append(time_array[-1] + self.settle_time)
+            velocity_array.append(self.v2)
+        time_array = np.around(time_array, 10)
+
+        # some of the math results in tiny fractions which affect some of the
+        # tests - round to 10 decimals
+        velocity_array = np.around(velocity_array, 10)
```

Comment:
Is this actually safe to do? We never store the absolute position, so rounding the velocities and times might change the end position...

Reply:
I think only by approx 10**-10. Is that an issue?

Reply:
I changed it to 12 decimals